<!-- mcl.launch -->
<launch >
    <arg name="node_name" default="mcl" />

    <!-- This is used for the test using mcl_3dl log data -->
    <arg name="use_sensor_points_merger" default="false" />

    <!-- yaml file of a target environment -->
    <!-- <arg name="map_yaml_file" default="$(find mcl3d_ros)/data/dist_map_hdl.yaml" /> -->
    <arg name="map_yaml_file" default="/media/akai/ssd/mcl3d_maps/kitti/05/dist_map_static_objects.yaml" />

    <!-- Initial pose (default initial pose is set for hdl log data) -->
    <rosparam param="/$(arg node_name)/initial_pose" > [0.0, 0.0, 0.0, 0.0, 0.0, 90.0] </rosparam >


    <!-- 0: measurement model optimization (similar to ICP scan matching) -->
    <!-- 1: particle filter (it does not work usually if odometry and/or IMU are not available) -->
    <!-- 2: fusion of particle-filter and optimization-based localizations -->
    <!-- 3: extended kalman filter -->
    <arg name="localization_mode" default="2" />

    <!-- 0: directory use residual errors (need to carefully set opt_max_error and this must be positive) -->
    <!-- 1: normal distribution (opt_max_error must be set from 0 to 1 if the measurement model type is 1, 2, and 3) -->
    <!-- 2: likelihood field model -->
    <!-- 3: class conditional measurement model -->
    <arg name="measurement_model_type" default="3" />

    <!-- TF frame names -->
    <arg name="map_frame" default="map" />
    <arg name="odom_frame" default="odom" />
    <arg name="base_link_frame" default="base_link" />
    <arg name="laser_frame" default="velodyne" />
    <arg name="opt_pose_frame" default="opt_pose" />

    <!-- input message names -->
    <arg name="sensor_points_name" default="/velodyne_points" />
    <arg name="odom_name" default="/odom" />
    <arg name="imu_name" default="/gpsimu_driver/imu_data" />

    <!-- use INS or not -->
    <arg name="use_odom" default="false" />
    <arg name="use_imu" default="false" />

    <!-- output msaage names -->
    <arg name="particles_name" default="/particles" />
    <arg name="opt_particles_name" default="/optimized_particles" />
    <arg name="pose_name" default="/mcl_pose" />
    <arg name="opt_pose_name" default="/opt_pose" />

    <!-- IMU frequency. This must be appropriately set if your IMU node does not output orientation. -->
    <arg name="imu_sample_freq" default="200.0" />

    <!-- resolution for voxel grid filter that will be applied to the sensor points before the scan registration -->
    <arg name="voxel_leaf_size" default="1.0" />

    <!-- number of particles -->
    <arg name="particle_num" default="1000" />

    <!-- number of sensor points used for likelihood calculation (this is used only when localization mode is 1)-->
    <arg name="sensor_points_num" default="1000" />

    <!-- measurement model parameters -->
    <arg name="z_hit" default="0.9" />
    <arg name="z_rand" default="0.05" />
    <arg name="z_max" default="0.05" />
    <arg name="var_hit" default="0.4" />
    <arg name="range_reso" default="0.1" />
    <arg name="range_max" default="120.0" />
    <arg name="unknown_lambda" default="0.001" />
    <!-- range_max and unknown_lambda must be appropriately set -->
    <!-- e.g., unknown_lambda is 0.001 if range_max is 120 m, and unknown_lambda is 0.01 if range_max is 30 m -->

    <!-- transformation between base link to laser -->
    <rosparam param="/$(arg node_name)/base_link_2_laser" > [0.0, 0.0, 0.0, 0.0, 0.0, 0.0] </rosparam >

    <!-- mcl3d node will broadcast tf if true -->
    <arg name="broadcast_tf" default="true" />

    <!-- make tf as map -> odom -> base_link -> laser if true, otherwise map -> base_link -> laser -->
    <arg name="use_odom_tf" default="false" />

    <!-- initial pose noise -->
    <rosparam param="/$(arg node_name)/initial_noise" > [0.1, 0.1, 0.1, 0.05, 0.05, 0.05] </rosparam >

    <!-- motion noise parameters -->
    <rosparam param="/$(arg node_name)/odom_noise" >
        [0.5, 0.01, 0.01, 0.01, 0.01, 0.01,
         0.01, 0.5, 0.01, 0.01, 0.01, 0.01,
         0.01, 0.01, 0.5, 0.01, 0.01, 0.01,
         0.01, 0.01, 0.01, 0.5, 0.01, 0.01,
         0.01, 0.01, 0.01, 0.01, 0.5, 0.01,
         0.01, 0.01, 0.01, 0.01, 0.01, 0.5]
    </rosparam >

    <!-- use linear interpolation to predict particle poses if true -->
    <arg name="use_linear_interpolation" default="true" />

    <!-- random particle rate. random particles are injected into re-sample phase if this value is larger than 0 -->
    <arg name="random_particle_rate" default="0.1" />

    <!-- re-sample threshold set to effective sample size -->
    <arg name="resample_threshold" default="0.5" />
    <rosparam param="/$(arg node_name)/resample_noise" > [0.1, 0.1, 0.1, 0.05, 0.05, 0.05] </rosparam >

    <!-- max iteration number for measurement model optimization -->
    <arg name="opt_max_iter_num" default="30" />

    <!-- error threshold for correspondences -->
    <arg name="opt_max_error" default="0.5" />

    <!-- convergence threshold set to errors -->
    <arg name="convergence_threshold" default="0.02" />

    <!-- number of particles sampled from the measurement model -->
    <arg name="optimized_particle_num" default="1000" />

    <!-- scale factor for covariance against an optimization result -->
    <arg name="optimized_pose_cov_coef" default="1.0" />

    <!-- GMM parameters used for likelihood calculation -->
    <arg name="gmm_postion_var" default="0.3" />
    <arg name="gmm_angle_var" default="0.1" />

    <!-- parameters for IMU update -->
    <arg name="ahrs_filter_kp" default="0.0" />
    <arg name="ahrs_filter_ki" default="0.0" />

    <!-- accept initialpose callback from rviz if true -->
    <arg name="use_initial_pose_cb" default="true" />

    <!-- write estimation log if true -->
    <!-- time mclx mcly mclz mclroll mclpitch mclyaw optx opty optz optroll optpitch optyaw -->
    <!-- if localization mode is 0, mcl and opt are the same -->
    <!-- if localization mode is 1, opt are zero -->
    <!-- if localization mode is 2, mcl and opt are different -->
    <!-- if localization mode is 3, mcl and opt are different and mcl represents ekf-based fusion result -->
    <arg name="log_file" default="/tmp/mcl3d_log.txt" />
    <arg name="write_log" default="true" />





    <!-- nodes -->
    <!-- Static TF from base link to laser -->
    <node name="base_link_to_laser" pkg="tf" type="static_transform_publisher" args="0 0 0 0 0 0 $(arg base_link_frame) $(arg laser_frame) 100" />

    <!-- This node marges several point clouds and outputs one point cloud. -->
    <node name="sensor_points_merger" pkg="mcl3d_ros" type="sensor_points_merger" if="$(arg use_sensor_points_merger)"/>

    <!-- Localizer -->
    <node name="$(arg node_name)" pkg="mcl3d_ros" type="mcl" output="screen" >
        <param name="map_yaml_file" value="$(arg map_yaml_file)" type="string" />

        <param name="sensor_points_name" value="$(arg sensor_points_name)" type="string" />
        <param name="odom_name" value="$(arg odom_name)" type="string" />
        <param name="imu_name" value="$(arg imu_name)" type="string" />

        <param name="particles_name" value="$(arg particles_name)" type="string" />
        <param name="opt_particles_name" value="$(arg opt_particles_name)" type="string" />
        <param name="pose_name" value="$(arg pose_name)" type="string" />
        <param name="opt_pose_name" value="$(arg opt_pose_name)" type="string" />

        <param name="map_frame" value="$(arg map_frame)" type="string" />
        <param name="odom_frame" value="$(arg odom_frame)" type="string" />
        <param name="base_link_frame" value="$(arg base_link_frame)" type="string" />
        <param name="laser_frame" value="$(arg laser_frame)" type="string" />
        <param name="opt_pose_frame" value="$(arg opt_pose_frame)" type="string" />

        <param name="use_odom" value="$(arg use_odom)" type="bool" />
        <param name="use_imu" value="$(arg use_imu)" type="bool" />

        <param name="broadcast_tf" value="$(arg broadcast_tf)" type="bool" />
        <param name="use_odom_tf" value="$(arg use_odom_tf)" type="bool" />
        <param name="use_initial_pose_cb" value="$(arg use_initial_pose_cb)" type="bool" />

        <param name="use_linear_interpolation" value="$(arg use_linear_interpolation)" type="bool" />

        <param name="imu_sample_freq" value="$(arg imu_sample_freq)" type="double" />

        <param name="localization_mode" value="$(arg localization_mode)" type="int" />
        <param name="measurement_model_type" value="$(arg measurement_model_type)" type="int" />

        <param name="particle_num" value="$(arg particle_num)" type="int" />

        <param name="sensor_points_num" value="$(arg sensor_points_num)" type="int" />

        <param name="random_particle_rate" value="$(arg random_particle_rate)" type="double" />
        <param name="resample_threshold" value="$(arg resample_threshold)" type="double" />

        <param name="z_hit" value="$(arg z_hit)" type="double" />
        <param name="z_rand" value="$(arg z_rand)" type="double" />
        <param name="z_max" value="$(arg z_max)" type="double" />
        <param name="var_hit" value="$(arg var_hit)" type="double" />
        <param name="range_reso" value="$(arg range_reso)" type="double" />
        <param name="unknown_lambda" value="$(arg unknown_lambda)" type="double" />
        <param name="range_max" value="$(arg range_max)" type="double" />

        <param name="voxel_leaf_size" value="$(arg voxel_leaf_size)" type="double" />

        <param name="opt_max_iter_num" value="$(arg opt_max_iter_num)" type="int" />
        <param name="opt_max_error" value="$(arg opt_max_error)" type="double" />
        <param name="convergence_threshold" value="$(arg convergence_threshold)" type="double" />

        <param name="optimized_particle_num" value="$(arg optimized_particle_num)" type="int" />
        <param name="optimized_pose_cov_coef" value="$(arg optimized_pose_cov_coef)" type="double" />
        <param name="gmm_postion_var" value="$(arg gmm_postion_var)" type="double" />
        <param name="gmm_angle_var" value="$(arg gmm_angle_var)" type="double" />

        <param name="ahrs_filter_kp" value="$(arg ahrs_filter_kp)" type="double" />
        <param name="ahrs_filter_ki" value="$(arg ahrs_filter_ki)" type="double" />

        <param name="log_file" value="$(arg log_file)" type="string" />
        <param name="write_log" value="$(arg write_log)" type="bool" />
    </node >
</launch >
